cputest/unaligned_stores.cpp

```diff
@@ -0,0 +1,204 @@
+#include <algorithm>
+#include <cinttypes>
+
+#include <gctypes.h>
+#include <ogc/irq.h>
+#include <wiiuse/wpad.h>
+
+#include "common/BitUtils.h"
+#include "common/hwtests.h"
+
+// This test covers the behavior described in https://bugs.dolphin-emu.org/issues/12565
+
+constexpr size_t PAGE_SIZE = 4096;
+constexpr uintptr_t CACHED_MEMORY_FLAG = 0x40000000;
+
+static volatile u32* const s_pi_reg = reinterpret_cast<volatile u32*>(0xCC003000);
+
+static volatile bool s_pi_error_occurred = false;
+
+static u32 Read(volatile u8* ptr)
+{
+  return *reinterpret_cast<volatile u32*>(ptr);
+}
+
+static void Write(volatile u8* ptr, u32 value, u32 size)
+{
+  switch (size)
+  {
+  case 4:
+    asm volatile("stw %0, 0(%1)" :: "r"(value), "r"(ptr));
+    break;
+  case 2:
+    asm volatile("sth %0, 0(%1)" :: "r"(value), "r"(ptr));
+    break;
+  case 1:
+    asm volatile("stb %0, 0(%1)" :: "r"(value), "r"(ptr));
+    break;
+  }
+}
+
+static u32 RotateRight(u32 value, u32 rotation)
```

Comment:
Move to bitutils.h?

Reply:
It turns out we already had an implementation of this in BitUtils.h and I just wasn't aware of it. I've made it use it now.